cmd/syncer/main.go

```diff
@@ -156,43 +170,93 @@ func startSyncers(ctx context.Context, GVRs []string, secretRef client.ObjectKey
 	specSyncConfig := syncer.Config{
-		UpstreamClientConfig: upstreamRestConfig,
-		GVRs:                 GVRs,
-		DownStreamClient:     mgr.GetClient(),
-		InformerFactory:      informerFactory,
-		ClusterID:            upstreamClusterConfig.Name,
-		NeverSyncedGVRs:      NEVER_SYNCED_GVRs,
+		GVRs:            GVRs,
+		InformerFactory: informerFactory,
+		ClusterID:       upstreamClusterConfig.Name,
+		NeverSyncedGVRs: NEVER_SYNCED_GVRs,
+		UpstreamNS:      controlPlaneNS,
+		DownstreamNS:    dataPlaneNamespace,
 	}
 
 	downstreamDynamicClient, err := dynamic.NewForConfig(mgr.GetConfig())
 	if err != nil {
 		return nil, fmt.Errorf("could not make dynamic downstream client from manager rest config: %v", err.Error())
 	}
-	SpecSyncer, err := spec.NewSpecSyncer(upstreamClusterConfig.Name, upstreamClusterConfig.Name, upstreamClient, downstreamDynamicClient)
+	SpecSyncer, err := spec.NewSpecSyncer(upstreamClusterConfig.Name, upstreamClusterConfig.Name, upstreamClient, downstreamDynamicClient, specSyncConfig)
 	if err != nil {
 		return nil, fmt.Errorf("could not create new spec syncer: %v", err.Error())
 	}
-	logger.Info("starting syncer", "name", upstreamClusterConfig.Name)
+	logger.Info("starting spec syncer", "name", upstreamClusterConfig.Name)
 
 	go SpecSyncer.Start(ctx)
 
-	syncRunnable := syncer.GetSyncerRunnable(ctx, specSyncConfig, syncer.InformerForGVR, SpecSyncer)
+	syncRunnable := syncer.GetSyncerRunnable(specSyncConfig, syncer.InformerForGVR, SpecSyncer)
+	return syncRunnable, nil
+
+}
+
+func startStatusSyncers(ctx context.Context, GVRs []string, secretRef client.ObjectKey, mgr manager.Manager, controlPlaneNS string) (*syncer.SyncRunnable, error) {
+	logger := log.FromContext(ctx)
+	controlPlaneSecret := &corev1.Secret{}
+	secretClient, err := client.New(mgr.GetConfig(), client.Options{})
+	if err != nil {
+		setupLog.Error(err, "unable to create sync secret client")
+		os.Exit(1)
+	}
+	err = secretClient.Get(ctx, secretRef, controlPlaneSecret, &client.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve control plane config secret '%v', error: %v", secretRef.String(), err)
+
+	}
+	downstreamClusterConfig, err := clusterSecret.ClusterConfigFromSecret(controlPlaneSecret)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create cluster config from control plane secret: %v", err)
+	}
+	downstreamClient, err := clusterSecret.DynamicClientsetFromSecret(controlPlaneSecret)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create client from control plane rest config: %v", err)
+	}
+	upstreamClient, err := dynamic.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, fmt.Errorf("unable to create client from mgr rest config: %v", err)
+	}
+	informerFactory := dynamicinformer.NewDynamicSharedInformerFactory(upstreamClient, DEFAULT_RESYNC)
+	informerFactory.Start(ctx.Done())
+	informerFactory.WaitForCacheSync(ctx.Done())
+
+	//for status: downstream is control-plane, upstream is data plane
```

Comment:
ahh. This is very confusing, is not simpler to think of CP as always being upstream and DP always being downstream?

Reply:
I intentionally avoided that, to keep the logic in the syncer from being so hardwired to our solution. Instead keeping the concept of upstream / downstream == from / to.